### ISSUE TITLE
Process labels

### ISSUE THREAD

**Issue by askbulle** [equinor/MAD-VSM-WEB#159]
**User story**
As an end user, I want to tag a process with keywords so that I can look up all processes with a common keyword

**Sharepoint ID**
34, 76, 92

**Describe the solution you'd like**
- Tags visible in overview (similar to Planner labels?)
- Examples: Asset (Åsgard, Snorre etc), Process (Well construction, department meeting etc)
- Examples: Project (Lean DPO)
- Pre-defined vs free text? If free-text, propose existing similar if creating new
- Project (list) - high potential - (#168 ) and Obsolete #173

**Additional context**
https://miro.com/app/board/o9J_lDp-Z84=/

---

### Requirements (for @ntorsen to test)
**Adding and removing labels on a process (Canvas)**
- [x] Every user can create and read labels.
- [x] I can't add labels to a process I don't have write-access to.
- [x] When adding a label, it should suggest an existing one. 

**Filtering processes on labels (front-page)**
- [x] It should be easy to find processes with a specific label.
- [x] Filter on one label
- [x] Filter on multiple labels. **It should follow _AND_ logic. That means all labels that is in the filter needs to be in a proccess for it to show.**
- [x] When choosing labels to filter with, I should be able to search for a label.
- [x] The labels shown in the label-view should be clickable (toggle active-state) and indicate that it is selected.
- [x] It should be easy to set and **clear** my label-filter.
- [x] I should be able to bookmark the current page (URL) for a specific search.

### Back-end
- [x] Create labels
- [x] Read labels
- [x] Get all labels (or a search endpoint?) 
- [x] Update text of a label (Superuser only) 
- [x] Delete an unlinked label (Superuser only)
- [x] Merge multiple labels into one (Superuser only)
- [x] ~~Return LinkCount in label response.~~
- [x] Add label filter parameter to process listings.
- [x] _Superuser_ should be able to administer labels for all processes
- [x] Filter Processes using AND logic (as explained above)

### Front-end
- [x] Update front-page 
![image](https://user-images.githubusercontent.com/3164065/136755125-bbbcc8ad-91fd-4b6e-9f24-3e502795b0de.png)
- [x] Create and add labels to a process (In the canvas) 
![image](https://user-images.githubusercontent.com/3164065/136909833-433b9cc0-d9f8-44da-91dc-060370669d2f.png)
![image](https://user-images.githubusercontent.com/3164065/135838539-bac75a96-aad7-462f-8f2d-a080f92b71b5.png)
https://www.figma.com/file/IkHwmIQrsT0iR34f5R5UnZ/VSM?node-id=2888%3A109052
- [x] Search function for labels.
- [x] Filter projects by labels.
- [x] Remove labels from a process.
- [x] When adding a label, get tips on an existing label to add. (Might need some design input on this ... 💅 )
- [x] #218 
- [x] #219

**Comment by SjoenH** [comment]
@Linedueo I think this one would be quite easy to do. Do you have a design on it?

**Comment by SjoenH** [comment]
<img width="966" alt="Skjermbilde 2021-09-26 kl  20 40 42" src="https://user-images.githubusercontent.com/3164065/134862239-a892cbe7-e5d3-4561-af9f-048872f80855.png">
Design for adding tags to a project. (in the canvas down right)
Still need design for showing tags in the front-page ( ref _**Tags visible in overview (similar to Planner labels?)**_) and possible to filter on tags. @Linedueo

**Comment by SjoenH** [comment]
Maybe we should call this **labels** instead of **tags**...

**Comment by SjoenH** [comment]
Do we need a label overview/edit page like in GitHub?

**Comment by SjoenH** [comment]
Can everyone create new labels? Should these labels be used as suggestions for other users? What happens then if someone creates a label for every letter in the alphabet?

**Comment by SjoenH** [comment]
@Linedueo could you talk with @knudsvik and @askbulle about the scope and design of this user story.

**Comment by knudsvik** [comment]
Sure give me a call @SjoenH

**Comment by SjoenH** [comment]
I had a call with @knudsvik 
On the points mentioned above, we don't want to govern the use of labels if we can help it. Let's see how it works out without limiting rubbish labels. _So **the issues I brought up** is a **non-concern** at the moment._

I believe these to be our immediate requirements:
- Every user can create and read labels
- You need access to the project to assign a label to it. 

Regarding #168 and #173 
With this solution, we could have labels for `Hight Improvement Potential` and `Obsolete`, but they would need to be added by the owner/editor of each project.
In the future we could add suport for a super-user to support adding labels to all projects if that would be helpful.

| Operation  | Future super-user | normal user |comment|
| ---------- | ------------- |--------|------|
| Create      | x  |x||
| Read         | x  |x||
| Update     | x  |-||
| Delete      | x  |-| Only accept delete if label is not assigned to any project.|

**Comment by SjoenH** [comment]
Status:
We are still waiting on design for labels on the front-page. @Linedueo do you have a status update on the progress of this?

Design for adding and removing labels to a project is ready. So moving this task to ready for development.

**Comment by Linedueo** [comment]
Design for labels on frontpage is being worked on- will update you asap

**Comment by Linedueo** [comment]
<img width="1514" alt="Skjermbilde 2021-10-04 kl  12 56 43" src="https://user-images.githubusercontent.com/74361011/135839709-6b384199-4ccd-4cb0-8471-d930b2ec5692.png">
@SjoenH @knudsvik

**Comment by Linedueo** [comment]
Figma link: https://www.figma.com/file/IkHwmIQrsT0iR34f5R5UnZ/VSM?node-id=2816%3A107643

**Comment by SjoenH** [comment]
Looks good to me. 👍
Does the cards have a fixed size or do they expand when adding labels￼?

Also, how do we filter on labels? 🏷

**Comment by Linedueo** [comment]
@SjoenH they expand when adding labels

**Comment by Linedueo** [comment]
@SjoenH Im not sure what would be a max amount of labels. But to not go overboard on adding labels we should probably have a maximum of 10? the cards will expand when adding labels so no fixed size.

**Comment by SjoenH** [comment]
@martabenthem You may start on the redesign of the `VSMCards`. #219

**Comment by Linedueo** [comment]
<img width="327" alt="Skjermbilde 2021-10-08 kl  08 33 26" src="https://user-images.githubusercontent.com/74361011/136509387-b3cbcd66-aac5-4513-91e2-132b6fc6c677.png">
@SjoenH updated design for adding labels- with suggestion when adding a new label

**Comment by Linedueo** [comment]
@knudsvik I will take a look at it.  I agree that it would be nice to see the most used labels

**Comment by Linedueo** [comment]
@knudsvik what do you think? The selected labels would show after closing so you can still refer to the ones you chose and also remove them. 
![   ](https://user-images.githubusercontent.com/74361011/136712615-fe0a9621-8db6-4d2b-8848-0bd53f5e5a5d.png)

**Comment by knudsvik** [comment]
Love it! 👍🏻

**Comment by SjoenH** [comment]
@Magos suggests that we use the sortLinkCount param when fetching the labels in the add-view. 
We are also considering adding that linkCount as a field in the label-object-response. I think it is a good idea, but only if it is quick to do...

**Comment by Linedueo** [comment]
![labels](https://user-images.githubusercontent.com/74361011/136909435-363a26c0-085d-494a-82ae-149b358d1ed6.png)

Figma 👉🏻      https://www.figma.com/file/IkHwmIQrsT0iR34f5R5UnZ/VSM?node-id=2888%3A109052

**Comment by martabenthem** [comment]
@Linedueo will there be a smartphone (small screen) version of the design for this feature?

**Comment by Linedueo** [comment]
@martabenthem  If it´s needed then yes. @knudsvik @askbulle is this something we should get done? If so, should it be prioritized?

**Comment by knudsvik** [comment]
We did agree that for all new features/designs we were to plan for mobile friendly layout from now on.

**Comment by martabenthem** [comment]
Hello! Should the labels be case-sensitive? 

If case-sensitive, we risk having "Kårstø" and "kårstø", which can be fixed with the merge functionality but it's not very valuable to be monitoring caps on labels.

If not case-sensitive, should we store labels as case-insensitive, and then show them with the first letter capitalised? Or show them with no caps at all ("kårstø")?

**Comment by knudsvik** [comment]
Good questions @martabenthem . I think they should not be case sensitive to avoid `Kårstø `and `kårstø `being two different labels.

**Comment by Magos** [comment]
Current status is that both searchText="alf" and searchtext="aLf" find the label "Svartalfrheim", so searching for an existing similar label is not case sensitive. It should be possible in the current system to make two labels that differ only in capitalization.

**Comment by SjoenH** [comment]
To-do:
- [ ] superuser access (not a requirement for this issue)

Labels are free-text input. That means it's up to the user to decide on the CapItaliZation and speling. 
But they will get previously used labels as a suggestion

**Comment by SjoenH** [comment]
Line will take a look at improving process-label edit and search for small screens.

**Comment by SjoenH** [comment]
@askbulle @knudsvik 
I told you yesterday that I thought this was going to be ready this week. But I forgot that @martabenthem is out of office and not working full-time on Flyt this week. Sorry about that.

We are currently looking at estimated release next week for this feature. We will keep you posted.

**Comment by martabenthem** [comment]
@ntorsen this feature is now ready for testing in the TEST environment :)

**Comment by martabenthem** [comment]
@Linedueo could you also take a look at the test environment and check this feature?

**Comment by ntorsen** [comment]
Tested all of the requirements (on top page ⬆️ ) and they worked good 👍 

Here are some findings I made when testing:
* Added a lot of process labels, and after doing that Filter by label does not look too happy. Maybe we can change the format here so that a user is able to scroll down a long list of labels?
![image](https://user-images.githubusercontent.com/89014017/145835315-2f634dfd-b39d-4eb8-8c29-1ce68146324c.png)

* There is a limit of 200 characters when adding process labels. When adding a label with 201 or more characters nothing happens after clicking on the Add-button. Suggestion: Can we add a message when a user tries to put in more than 200 characters, so that he/she is aware that this is a restraint? 

* Tried putting in an emoji when adding label, but then I get the following:
![image](https://user-images.githubusercontent.com/89014017/145834942-d2382f51-6c08-4c5b-b98a-a0d177f52a38.png)
![image](https://user-images.githubusercontent.com/89014017/145834997-dc4a5f96-d374-4ca1-b86f-bd21ca633a51.png)

The first bullet point I find most important. The following two is not blockers, but nice to have fixed. 
@martabenthem @SjoenH

**Comment by ntorsen** [comment]
**Test of new updated version:**
- Filter by label is looking good 😊
- User gets a message when typing in more than 200 characters 🥇

![image](https://user-images.githubusercontent.com/89014017/146005511-51f9eeab-2c84-48a8-bbb2-b5cba077dcbc.png)
- Found out that some emojis work, while others don't. But I agree with you @martabenthem, we can see if this becomes a concern

**Major difference from before;** User cannot add label when clicking on a specific label in the process from landing page/see my processes/see my favorite processes. If you click it now, it filters by that label. Before you could add directly from that page without having to go into the process. 

![image](https://user-images.githubusercontent.com/89014017/146006802-bf24b782-c8df-4a1e-acae-89ec12436479.png)

**Comment by SjoenH** [comment]
@ntorsen You can add labels by clicking on the label-icon.

**Comment by martabenthem** [comment]
@ntorsen sorry, I forgot to mention that change in functionality. 

- I added a "tag" button at the bottom of process cards. It only appers in processes where you have edit access.
- You can click on any label and that will trigger filtering by that label.

**Comment by SjoenH** [comment]
![image](https://user-images.githubusercontent.com/3164065/146017109-ca226660-098a-4ac5-b9e0-f2e6468fd052.png)
@martabenthem I still get these inactive scroll-bars.

**Comment by martabenthem** [comment]
@SjoenH do you get inactive scroll-bars when you open the ManageLabelBox too?

**Comment by knudsvik** [comment]
When I have this dialogue box open and scroll with the mouse wheel, the background is scrolled, not the dialogue box. Similar strange behaviour appears in the dialogue box when opening from the process

![image](https://user-images.githubusercontent.com/25986427/146026866-fcd11b04-b260-4216-8f2e-af08c9038cbf.png)

**Comment by martabenthem** [comment]
@knudsvik thank you for the feedback. Should be fixed now.

**Comment by knudsvik** [comment]
@SjoenH @martabenthem : We are contemplating (again) whether making new labels should be open for the users. Yes we know that we discussed this and agreed upon having it open a long while ago 🤬. The risk of releasing it with the intended functionality is that it rapidly grows out of control and loses value. Changing it to a locked list on the other hand; not very nice to you developers and a maintenance risk for us (and maybe we eventually want to have it open anyway)... 

For us to make a more informed decision: How will a change like this affect the coding process now? How big change is this?

**Comment by SjoenH** [comment]
@knudsvik it's not that big of a change to disable it in the front-end... But that is just safety by obscurity and it would be best to limit that interaction in the back-end as well. But it might be good enough for now? :man_shrugging: 
Not sure how long time it will take to fix in the back-end. @Magos 

**A note about managing labels.**
@Magos made an endpoint to ease combining multiple labels into one. 
![image](https://user-images.githubusercontent.com/3164065/146049239-024eb5f0-cedf-4b89-b510-4bef47fb6c5f.png)
Let's say we have 10 labels with different spelling, then we can fix them all in one request.

Let us know what you think.. I'm also available to chat about it tomorrow :grin:

**Comment by SjoenH** [comment]
Btw. I created a label for our team and this feature makes it so much easier to share all our processes. :100:

**Comment by knudsvik** [comment]
@SjoenH disabling it in the front-end may be a good solution for initial release. Maybe we could revisit after a month and decide to either disable back-end or opening it up? That way we have a better feeling about it after some initial use and can get some feedback. 

If so it would be good to have the import tool you screenshot on Slack available in the interim period. That way we can add more labels by request in the start.

**Comment by SjoenH** [comment]
Sounds like a plan :muscle: I'll spend some time tomorrow on improving label management. Nothing fancy. Design by Henry. :nerd_face:

**Comment by knudsvik** [comment]
Doesn't have to be fancy, it's not for the customers ;)

**Comment by Magos** [comment]
> @knudsvik it's not that big of a change to disable it in the front-end... But that is just safety by obscurity and it would be best to limit that interaction in the back-end as well. But it might be good enough for now? 🤷‍♂️ Not sure how long time it will take to fix in the back-end. @Magos

I suppose the main option right now would be to introduce a check for Super-user access before you're allowed to create a new label text. The new label route and project->Add label route would both need to introduce this check if the label text is not an existing one.

**Comment by SjoenH** [comment]
Todo:
- [ ] We should only be able to add existing labels to a process.

Frontend:  

@martabenthem 
- [ ] Remove the add button and enter-key listener
- [ ] Green outline when active (EDS Design)
- [ ] Do not round the corners (EDS Design)
- [ ] Grey background in the input field (EDS Design)
- [ ] Autofocus (check if it works fine)

@SjoenH 
- [ ] Admin labels page

Back-end: @Magos 
- [ ] user needs to be super-user to create new labels

**Comment by martabenthem** [comment]
@Magos FYI, my changes are now deployed in TEST.

**Comment by Magos** [comment]
My changes to the backend [are now deploying to TEST environment also](https://dev.azure.com/statoil-mad/VSM/_releaseProgress?_a=release-pipeline-progress&releaseId=229).

**Comment by SjoenH** [comment]
Ready for testing in Test & QA @knudsvik @askbulle @ntorsen 

Label admin can be reached at:  
- https://web-vsm-test.radix.equinor.com/labels/bulkAdd
- https://web-vsm-qa.radix.equinor.com/labels/bulkAdd

Note: Only superUsers can create labels. (@ntorsen I'm sorry. But you may verify that you cannot create any :angel: )

**Comment by knudsvik** [comment]
@SjoenH : ### Result from QA

The label sorting, filtering etc functionality works well, also tried the bulk importer which seems good for initial usage. Adding labels both from the front page and the canvas is missing something in our minds. 

![image](https://user-images.githubusercontent.com/25986427/146214394-f0cc315d-4863-4b74-aff7-874ad15aa0e5.png)

The user should ideally see the available labels and choose from them (and search to narrow the amount), much like the filtering window looks like. 

![image](https://user-images.githubusercontent.com/25986427/146214917-98f23b00-fd96-4c81-9e47-61086ded1e75.png)

Maybe the ones active for the process could be highlighted and listed on the top? Please let me know what you think and if you want to discuss this.

**Comment by ntorsen** [comment]
I can verify that I'm not able to add any process labels (since not superUser) ✔

**Comment by martabenthem** [comment]
@Linedueo what do you think about the feedback above?
It could be useful to have an overview of available labels as suggested by @knudsvik, but it could also become difficult to actually read them when the number of labels increases.

**Comment by Linedueo** [comment]
@martabenthem Best solution I think would be to do what @knudsvik mentioned. The available labels should show and then you start filtering by typing. But we should probably have max number of labels shown. I would think a maximum of 3-5 lines  of labels and then scrolling if there are more.

**Comment by martabenthem** [comment]
![image](https://user-images.githubusercontent.com/61791419/146745668-98322171-e2b8-497a-9276-5e7c868d1356.png)

**Comment by Linedueo** [comment]
@martabenthem Thorjan and I came up with a solution. We put a search bar on top, and below the search bar you have active labels followed by all labels. Then of course you can filter labels by searching. Let me know if you have questions.
<img width="988" alt="Skjermbilde 2021-12-20 kl  12 09 00" src="https://user-images.githubusercontent.com/74361011/146758408-5ee69554-f575-40d3-ae96-94b8155e41fd.png">

**Comment by martabenthem** [comment]
@Linedueo are the existing labels also filtered when you type on the search bar? Or do they stay always there irrespective of what you type in the search bar?

**Comment by knudsvik** [comment]
I think we should leave the existing lables there irrespective of the searching.. Can always change this later if it feels inconsistent?

**Comment by martabenthem** [comment]
@ntorsen the change discussed above is now in TEST. 
@Linedueo you might want to check TEST as well.

**Comment by ntorsen** [comment]
@martabenthem for transparency reasons I will also list my findings here. 
When adding a label to a process that has a long text inside, the bracket looks kind of "cut off" 
![image](https://user-images.githubusercontent.com/89014017/146917283-1bf524c3-f663-4ea1-be8c-ebafabb42ee5.png)

And when clicking on Filter by label I get two scroll bars. The first one seems inactive, its the second one I use to scroll down. 
![image](https://user-images.githubusercontent.com/89014017/146917554-57c99e42-c75b-4664-9dcd-57a8ab6af735.png)

**Comment by martabenthem** [comment]
@ntorsen the scroll issues should be fixed now. 

@knudsvik I had not thought of whether the user should be able to add labels both from current process and to-be process, what do you think?

**Comment by knudsvik** [comment]
> @knudsvik I had not thought of whether the user should be able to add labels both from current process and to-be process, what do you think?

It seems like the version in test now allows you to add labels from both Current and To-Be (and they are shared). I think that is how it supposed to be!

**Comment by martabenthem** [comment]
@knudsvik not exactly, I think for now adding from the To-Be process doesn't work. But I will look into implementing it such that adding to the Current also adds to the To-Be process, and vice-versa. Same for removing.

**Comment by knudsvik** [comment]
Ok, I am able to remove from both Current and To-Be, but only add from Current.

**Comment by martabenthem** [comment]
@Magos I think this is something that could potentially be handled in the backend - that adding an removing labels from the current process does the same operation to to To-Be and viceversa. What do you think?

**Comment by martabenthem** [comment]
@ntorsen @knudsvik I have addressed the scrolling issues. Labels are added/removed always from both the current and the to-be processes. These fixes are now deployed to TEST.

**Comment by ntorsen** [comment]
@martabenthem I still get the two scrollbars in Filter by label (opened up a incognito window for this new test).  

For the labels that contained a long text in a process they was sort of cut off before, now the whole text is visible with an added scrollbar below to view it. Is it possible to somehow make the labels look whole without having this scrollbar? Maybe with some dots (...) after a certain number of characters? Don't know what would be best design-wise, or if it is preferable as it is..

![image](https://user-images.githubusercontent.com/89014017/146937530-2a2fd20d-ab88-4386-9f7e-be13634920f3.png)

I am able to add/remove from the current and the to-be process now  💯

**Comment by martabenthem** [comment]
@ntorsen in my browser I don't see inactive scrollbars so I don't have a way of checking. I think now it should have been fixed.

@knudsvik do you have an opinion about the design of long labels on the process cards? I think this is an edge case since labels are rarely that long.

**Comment by knudsvik** [comment]
> @knudsvik do you have an opinion about the design of long labels on the process cards? I think this is an edge case since labels are rarely that long.

Yes, agree that it is an edge case. Not sure if it needs to be adressed now since we are also in control of the labels, will not make any this long. Maybe we could fix this if it becomes a problem later?

**Comment by ntorsen** [comment]
For Chrome and Edge it looks like this now in Filter by label: 

![image](https://user-images.githubusercontent.com/89014017/147064220-545936b5-bd8b-4e7f-a145-ffad17aa3350.png)

**Comment by martabenthem** [comment]
@knudsvik I think that makes sense. I will release to QA!

**Comment by martabenthem** [comment]
@ntorsen ready for a new round of testing!

**Comment by ntorsen** [comment]
Done another run-through of requirements of testing (checklist on top). Fixes that has been made; 
- Label text is showing whole text
- Filter by label: now has only one scroll bar and pop-up box looks right
- Able to add labels from current and to-be

**BUT only able to remove labels from current, not from to-be process** 
@martabenthem

**Comment by Magos** [comment]
Quick question: Is this "close" text intended? The button kinda more represents "remove" or something, I feel.
![u9polMfgK6](https://user-images.githubusercontent.com/1004386/147938455-7c1b106d-837d-4537-87b1-f6ba868bd40d.png)

**Comment by Magos** [comment]
> **BUT only able to remove labels from current, not from to-be process** @martabenthem

I checked the test environment database, and found some of the processes to be in an inconsistent state (label-links stored on the to-be process) that should not be possible to create after this latest backend fix. I've corrected those entries by deleting the labels from those processes.

**Comment by ntorsen** [comment]
Working good for test environment now 😊

**Comment by martabenthem** [comment]
Great! @knudsvik and @askbulle this feature is now deployed in QA.

**Comment by knudsvik** [comment]
### Results from QA:

We have discussed the sorting of labels in the pop-up boxes but landed on creating a new issue which could take care of other sorting options and releasing as is now as MVP. New issue created for sorting:
- #345 

Otherwise it looks good except for one thing. When we are going to type in the "Search labels" bar, internet explorer wants to propose stuff we have written earlier. This covers over the actual labels we want to choose and is incredibly irritating. Is it possible to force this behaviour off?
![image](https://user-images.githubusercontent.com/25986427/148543035-0844e3fb-b7e5-4630-a6eb-e39dbf730b35.png)

**Comment by SjoenH** [comment]
@knudsvik Super... I'll take a look at it :+1:

**Comment by SjoenH** [comment]
@knudsvik I didn't get it reproduced on my end, but I've released a new build that should technically disable it. 
Could you take a look in QA?

**Comment by knudsvik** [comment]
@SjoenH perfect now!

**Comment by SjoenH** [comment]
Released to Prod
